internal/controller/controllers/agentserviceconfig_controller.go

```diff
@@ -922,13 +950,34 @@ func (r *AgentServiceConfigReconciler) newAssistedServiceDeployment(log logrus.F
 		var replicas int32 = 1
 		deployment.Spec.Replicas = &replicas
 		deployment.Spec.Strategy = deploymentStrategy
+
+		// Handle our hashed configMap(s)
+		if deployment.Spec.Template.Annotations == nil {
+			deployment.Spec.Template.Annotations = make(map[string]string)
+		}
+		deployment.Spec.Template.Annotations[assistedConfigHashAnnotation] = assistedConfigHash
+		if mirrorConfigHash != "" {
+			deployment.Spec.Template.Annotations[mirrorConfigHashAnnotation] = mirrorConfigHash
+		}
+		if userConfigHash != "" {
+			runningUserConfigHash, ok := deployment.Spec.Template.Annotations[userConfigHashAnnotation]
+			if ok {
+				if runningUserConfigHash != userConfigHash {
+					msg := fmt.Sprintf("User provided and running configuration differ on assisted-service deployment. Remove the '%s' annotation from deployment to force restart", userConfigHashAnnotation)
+					r.Recorder.Event(instance, "Warning", "UserConfigMismatch", msg)
```

Comment:
If we're adding a hash of the configmap to the pod spec template, won't that automatically cause a new rollout? Or have we not fully implemented that?

Reply:
Yeah. This will cause a rollout, however we already do that when we update the env because the user supplied the "unsupported" configmap (https://github.com/openshift/assisted-service/pull/2137/files#diff-c619bb36181b5e1dd76fb1fcc1f54068b16a8761f77106052fcbf5e2f5f9d356R730-R737) quoting so it doesn't get lost:

```
	userConfigName, ok := instance.ObjectMeta.GetAnnotations()[configmapAnnotation]
	if ok {
		log.Infof("ConfigMap %s from namespace %s being used to configure assisted-service deployment", userConfigName, r.Namespace)
		...code omitted for space...
		envFrom = append(envFrom, []corev1.EnvFromSource{
			{
				ConfigMapRef: &corev1.ConfigMapEnvSource{
					LocalObjectReference: corev1.LocalObjectReference{
						Name: userConfigName,
					},
				},
			},
		}...,
		)
	}
```

Let me know if you see a fault in my line of thinking here.

Reply:
If you modify a Deployment and change anything about the pod template, the deployment controller will automatically roll out new pods to replace the old pods. It doesn't matter what the change was. Putting a hash in an annotation on the pod is a technique for using that behavior to our advantage. All we have to do is keep the hash current, and the Deployment controller will do a roll-out any time it changes.